### ISSUE TITLE
Clarification on estimators.ApplyExistingEstimates and parameters.SetValues

### ISSUE THREAD

**Issue by halabikeren** [veg/hyphy#753]
Dear HyPhy team,

What is the purpose of the function estimators.ApplyExistingEstimates()? Specifically, how is it different from the function parameters.SetValues()? Is there any way to recognize states where the usage of parameters.SetValues() is required rather than estimators.ApplyExistingEstimates(), based on the initial_values argument, given to the function calling any of these functions?

Thanks!
Keren

**Comment by halabikeren** [comment]
Hi,

I also noticed that assigning values to constrained parameters (as the function parameters.SetValues() does in my case), the outcome is removal of the constrain on the parameter.

For example, the following code segment :

```
DataSet data = ReadFromString (
"
>A
A
>B
C
>C
A
>D
A
"
);
treeStr = "(A:0.2,B:0.1,(C:0.2,D:0.2):0.2);";

DataSetFilter dataFilter = CreateFilter (data,1);

global mu;
mu := 42;
RateMatrix = {{,mu,mu,mu}
{mu,,mu,mu}
{mu,mu,,mu}
{mu,mu,mu,}};
Freqs = {{0.25},{0.25},{0.25},{0.25}};
Model M = (RateMatrix, Freqs);

Tree tree = treeStr;

LikelihoodFunction LF = (dataFilter, tree);

GetString(lf_info, LF, -1);
fprintf(stdout, "lf_info settings before setting value of mu:\n");
fprintf(stdout, "global independent parameters: ", lf_info["Global Independent"], "\n");
fprintf(stdout, "global constrained parameters: ", lf_info["Global Constrained"], "\n");  


mu = 1;  // set value

fprintf(stdout, "\n\n");
GetString(lf_info, LF, -1);
fprintf(stdout, "lf_info settings after setting value of mu:\n");
fprintf(stdout, "global independent parameters: ", lf_info["Global Independent"], "\n");
fprintf(stdout, "global constrained parameters: ", lf_info["Global Constrained"], "\n");  

```

Yields:

```
lf_info settings before setting value of mu:
global independent parameters: {

}
global constrained parameters: {
{"mu"}
}


lf_info settings after setting value of mu:
global independent parameters: {
{"mu"}
}
global constrained parameters: {

}

```

Why should an assignment of a value to a parameter result in the removal of its constraint?

Thanks again!
Keren

**Comment by spond** [comment]
Dear @halabikeren,

Assigning a value to a constrained variable **must** remove the constraint. 

```
x := expression;

// x is now a FUNCTION of other parameter values

x = value;

// x is now an unconstrained variable set to new value
// otherwise you have something like

expression = value;

// which is really doesn't make sense, because  you can't say

2+3x = value;

// unless you treat this as an equation to solve

```

In your example, you seem to wish to change the constraint on `mu`. Do so by replacing

```
mu = 1;  // set value
```

with

```
mu := 1;  // update constraint
```

Best,
Sergei

**Comment by halabikeren** [comment]
Thank you Sergei,

Sorry, this is a silly example. I should have clarified that when referring to removal of constraints, I was meaning any constraint on the parameter (and not necessarily a hard assignment on it, such as `mu := 1`). 

Here is a more relevant example:

```
DataSet data = ReadFromString (
"
>A
A
>B
C
>C
A
>D
A
"
);
treeStr = "(A:0.2,B:0.1,(C:0.2,D:0.2):0.2);";

DataSetFilter dataFilter = CreateFilter (data,1);

global a;
global b;
a := b;
RateMatrix = {{,a,a,a}
{a,,a,a}
{a,a,,a}
{a,a,b,}};
Freqs = {{0.25},{0.25},{0.25},{0.25}};
Model M = (RateMatrix, Freqs);

Tree tree = treeStr;

LikelihoodFunction LF = (dataFilter, tree);

GetString(lf_info, LF, -1);
fprintf(stdout, "lf_info settings before setting value:\n");
fprintf(stdout, "global independent parameters: ", lf_info["Global Independent"], "\n");
fprintf(stdout, "global constrained parameters: ", lf_info["Global Constrained"], "\n");  


a = 1;  // set value

fprintf(stdout, "\n\n");
GetString(lf_info, LF, -1);
fprintf(stdout, "lf_info settings after setting value:\n");
fprintf(stdout, "global independent parameters: ", lf_info["Global Independent"], "\n");
fprintf(stdout, "global constrained parameters: ", lf_info["Global Constrained"], "\n");  
```

This yields:

```
global independent parameters: {
{"b"}
}
global constrained parameters: {
{"a"}
}


lf_info settings after setting value:
global independent parameters: {
{"b", "a"}
}
global constrained parameters: {

}
```

In such case, why would this assignment result in the removal of the constraint that defines the dependency between parameters `a` and `b`?

In the context of using the function `parameters.SetValues()` on a model in which there are dependencies between parameters (such as RELAX), this phenomenon poses the need to re-apply the constraints defining the  dependencies between the parameters of the model.

Regarding the difference between `parameters.SetValues()` and `estimators.ApplyExistingEstimates()`:
I see that `parameters.SetValues()` also traverses over all the parameters in the provided dictionary and assigns the values in it to the parameters of the model. Thus, it seems that it can achieve the same purpose that use described above. So, I still don't quite understand how it is different from `estimators.ApplyExistingEstimates()`.

Thanks!
Keren

**Comment by spond** [comment]
Dear @halabikeren,

Imagine trying to write the following statements in Python

```

def x (a) {
    return 2*a
}

x = 3;

```

The behavior defined here is to replace the functional binding of `x` to an expression (function) with the new fixed value. 

This is exactly what happens with HyPhy when you write

```
a := b; // a is now an expression whose value is determined by b

a = 1; // a is now a value

```

If you don't do this, then you can easily arrive at logical nonsense

```
b = 1;
a := b;
a = 2; 
```

If `a` is now equal to 2, and the constraint is maintained, then `a` must also be `1`. Or we need to "solve" the expression for `b` and update its value to be consistent with the new value of `a`.

------

`estimators.ApplyExistingEstimates()` will apply estimates based on their annotated meanings (not their names). `SetValues` simply copies values to identifiers 

```
parameters.SetValues ({"x" : 2, "y": 3});
=>
x = 2;
y = 3;
```

With `estimators.ApplyExistingEstimates()` you get something different. For example your model can have a global parameter "kappa" whose attributed meaning is "Transitition/transversion ratio", e.g. 

```
"parameters":{
   "global":{
     "Transitition/transversion ratio" : "kappa",
...
...
}
 
```

Your parameter estimates may have a parameter named "ts_tv" whose attributed meaning is also "Transitition/transversion ratio", e.g. 

```
"Transitition/transversion ratio":{
     "ID":"ts_tv",
     "MLE":5.6119000113439
    }
```

`estimators.ApplyExistingEstimates()`  will copy the appropriate estimate (5.611...) into the variable "kappa".

`estimators.ApplyExistingEstimates()` can also selectively apply constraints when copying parameters, and **importantly** it copies branch lengths between **differrent** models, so they mean the same thing, say in nucleotide and codon spaces.

Best,
Sergei

**Comment by halabikeren** [comment]
Thank you for the clarification, Sergei,

I have noticed that if I assign a value to a parameter before setting a constraint on it, the constraint is retained, and the value is assigned.

I have also noticed that when a parameter is already constrained, `estimators.ApplyExistingEstimates() ` would not assign a value to it, even if a value is available in the input dictionary.

I wish to re-create a model and define its constraints iteratively, while assigning to its parameters values from the previous iteration (that used a different model instance).  In this iterative procedure, I fix the branch lengths, so I have no interest in assigning their corresponding local parameters values from the previous iteration. Thus, I could simply:
1) Create the model.
2) Assign values to its parameters.
3) Apply the constraints.

Using this order of actions would allow me to assign the values I wish to the parameters of the model while retaining their constraints.

**My question is:** In step 2, I could either use `parameters.SetValues()` or `estimators.ApplyExistingEstimates()`:
The initial values have the exact same names, so `parameters.SetValues()` could work.
The constraints are not applied yet, so `estimators.ApplyExistingEstimates()` could work.
Is there a difference in the outcomes (only on the global parameters of the model) of using each of the above functions in this case?

Thank you!
Keren
